### ISSUE TITLE
Add ListRoles function to Keystone v3

### ISSUE THREAD

**Issue by jamiehannaford** [gophercloud/gophercloud#550]
Currently we don't support the [list roles](https://developer.openstack.org/api-ref/identity/v3/#list-roles) function. Ideally a PR would implement the full CRUD set for Roles, but list would be a good first step.

**Comment by dklyle** [comment]
I can take this.

**Comment by dklyle** [comment]
Looking at the role content that exists around role assignments. I will likely want to refactor the exiting RoleAssignment struct as it includes structs for Role, User, and Group which are defined in role/results.go As there are structs defined elsewhere that match these already and I will need to define a Role struct for the context of listing roles that has more information. Just want to make sure that such a internal plumbing change is OK before moving forward. Thoughts?

**Comment by jtopjian** [comment]
Quite coincidentally, I'm working with the RoleAssignment API in Python at this moment.

From what I can see, the information contained in the result of a RoleAssignment entry is only a subset of a full domain, user, group:

```
<RoleAssignment 
  role={u'id': u'bdfee2f5869f43508a4f881164d04c16', u'name': u'Member'}, 
  scope={u'project': {u'domain': {u'id': u'default', u'name': u'Default'}, u'id': u'2f77533e2d6148189da24a3cbe707859', u'name': u'jdoe@gmail.com'}}, 
  user={u'domain': {u'id': u'default', u'name': u'Default'},  u'id': u'24cc044d675f45098afab4a0e3bb5b77', u'name': foobar'}>
```

This also seems to match the API docs: https://developer.openstack.org/api-ref/identity/v3/#id271

I haven't done a raw curl command to verify, though.

Since the `role_assignment` API is only returning a subset of information, I'm leaning on leaving the structs as they are. I get that they are _kind of_ duplicate, but also kind of not. I see them as two different types of result. 

If we were to swap these structs out with the more robust ones in other packages, IMO I think it would be more confusing to a end-user/dev to have a `projects.Project` result with half of the fields empty.

Unless I totally misunderstood what you were proposing :/

**Comment by dklyle** [comment]
My biggest concern is that there is a role struct that is defined as just an ID. A true Role struct with name, domain_id, ID, etc is more correct. I would prefer to at a minimum rename the existing Role struct to something else or I'll have to invent a less natural name to match what I would consider the full Role struct. The others (user, group) would just be potentially confusing.

**Comment by jtopjian** [comment]
> My biggest concern is that there is a role struct that is defined as just an ID. A true Role struct with name, domain_id, ID, etc is more correct.

Agreed. the `roles.Role` struct most definitely needs fleshed out.

> I would prefer to at a minimum rename the existing Role struct to something else

six-one-way-half-a-dozen-the-other :)

Either adding additional fields to the current `Role` struct and creating something like a `RoleAssignment` struct _or_ renaming the existing `Role` to `RoleAssignment` and creating a new `Role` struct with all known fields that a GET would return.

Whatever the names turn out to be, I do understand what you're saying here and I agree.

> The others (user, group) would just be potentially confusing.

Just to be overly verbose on this one, the `roles.User`, `roles.Project`, etc structs, while initially looking incorrect, are actually correct from the point of view of the `role_assignments` API call.

If it turns out there is _another_ role API call which would return the full amount of information for any of these, then either we can look at importing the `users.User`, `projects.Project` structs _or_ doing the same thing as renaming `roles.Role`. Adding on to the latter, renaming `roles.Project` to `roles.ProjectAssignment` (or something similar) might be a good long-term idea.

And for even more verbosity, I did a raw curl command, and can confirm that `role_assignments` is indeed not sending back the full amount of information, so we must adhere to this, in some way, in the `roles` package:

```shell
#!/bin/bash

oscurl() {                                     
  token=$(openstack token issue -c id -f value)
  curl -s -H "X-Auth-Token:$token" $1          
}                                              

oscurl http://localhost:35357/v3/role_assignments?include_names=True | python -mjson.tool
```

That yields output like the following:

```json
{
    "links": {
        "next": null,
        "previous": null,
        "self": "http://localhost:35357/v3/role_assignments?include_names=True"
    },
    "role_assignments": [
        {
            "links": {
                "assignment": "http://localhost:35357/v3/projects/9b0e16aeb6934f06ab1d87036052bbcb/users/130bd6dcc4014804beb25490b652dceb/roles/f3f979f71d924db69154ccdb4b9b5c03"
            },
            "role": {
                "id": "f3f979f71d924db69154ccdb4b9b5c03",
                "name": "admin"
            },
            "scope": {
                "project": {
                    "domain": {
                        "id": "default",
                        "name": "Default"
                    },
                    "id": "9b0e16aeb6934f06ab1d87036052bbcb",
                    "name": "services"
                }
            },
            "user": {
                "domain": {
                    "id": "default",
                    "name": "Default"
                },
                "id": "130bd6dcc4014804beb25490b652dceb",
                "name": "nova"
            }
        },
```

It looks like `Names` was introduced in Keystone 3.6. Given I can use `?include_names=True` in an Ocata environment without any microversion customization, I think we can add `Name` to the various `roles` struct... but that doesn't have to be immediately done.

**Comment by jtopjian** [comment]
Implemented in #561
